### ISSUE TITLE
Can we use DALI to build a data loader which uses an pre-trained inference model?

### ISSUE THREAD

**Issue by Rashindrie** [NVIDIA/DALI#2684]
Hi,

I was wondering if DALI is capable of being used to build a data loader which utilizes a pre-trained model for feature extraction.

To explain my use case: Given a set of images and a single label for those set of images, I want to first get feature representations for each image using a pre-trained model and then output those features as the output of the data loader.

Since feature extraction is done, the pre-trained model needs to be kept in GPU for efficient forward pass. The extracted features for all images in the set together with a overall label would be output as (image_feature_set, label).

I am aware that I can pre-compute these features and store them and then use the data loader to access the files directly thereby avoid using a GPU-model in the data loader.

However, that option is bit limiting as we wont be able to get augmentations on the fly.

Any insight is appreciated.

Thanks!

**Comment by JanuszL** [comment]
Hi,
As I understand you want to run the inference as a part of the DALI pipeline.
There was an idea to run TRT as a custom DALI operator, you can learn more about it here:
- GPU Technology Conference 2019; Integration of DALI with TensorRT on Xavier; Josh Park and Anurag Dixit: slides, recording.

The other idea that comes to my mind is to utilize the Python operator and just run the inference there.

**Comment by Rashindrie** [comment]
Hi @JanuszL,

Yes, that's what I want to do. 

I would like to integrate DALI with pytorch. I'm not sure I am clear on what you mean by utilizing a python operator.

In Pytorch, when you try to use num_workers > 1 with CUDA, there are multiprocessing errors (https://discuss.pytorch.org/t/model-inference-inside-custom-dataset-class/16640). 

Hence why I was thinking if DALI might be of assistance.

Thanks,
Rashindrie

**Comment by JanuszL** [comment]
What I meant is you can implement the DALI operator using plain Python code and run the inference part inside it. You can read more about it [here](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/custom_operations/python_operator.html).
However, you won't be able to utilize multithreading in this operator and it will be synchronous.

**Comment by Rashindrie** [comment]
Right, got it. I'll go through the references.

Thanks alot for your comments @JanuszL. Much appreciated :)
